### PR TITLE
afl++: increase forkserver timeout

### PR DESCRIPTION
**Pull request by vanhauser-thc**
This PR should first be tested with projects that make trouble if this really fixes it.
It increases the forkserver timeout plus removes dont_defer_forkserver.
also updates afl++ commit id and removes a command line option that is now default.

### CODE REVIEW

**Comment by vanhauser-thc** [comment]
I added instrumentation tuning to focus more on bug finding rather than finding new paths, and the new commit it fixed a bug when instrumenting tinyxml2.
@inferno-chromium  @jonathanmetzman  @MarcoFalke @catenacyber  can you have a look please?

**Comment by catenacyber** [comment]
I am not sure I understand. But the last commit `tune afl++ instrumentation, fix llvm dict2file bug` seems to just change probabilities.
If so, we still may trip on the bug, but it will just be less likely.
Am I missing something ?

**Comment by vanhauser-thc** [comment]
@catenacyber for this look up one commit above :)

**Comment by MarcoFalke** [comment]
Sorry, I am not sure how to test [infra] changes. The issue on current master is:


<details><summary>afl timeout on master</summary>

```
# python infra/helper.py check_build  --sanitizer address  --engine afl $PROJECT_NAME
Running: docker run --rm --privileged -i -e FUZZING_ENGINE=afl -e SANITIZER=address -e ARCHITECTURE=x86_64 -e FUZZING_LANGUAGE=c++ -v /root/oss-fuzz/build/out/bitcoin-core:/out -t gcr.io/oss-fuzz-base/base-runner test_all.py
INFO: performing bad build checks for /tmp/not-out/process_message_filterclear
INFO: performing bad build checks for /tmp/not-out/cryptofuzz-bitcoin-cryptography-w15-p4
INFO: performing bad build checks for /tmp/not-out/cryptofuzz-bitcoin-cryptography-w24-p8
INFO: performing bad build checks for /tmp/not-out/cryptofuzz-bitcoin-cryptography-w2-p2
INFO: performing bad build checks for /tmp/not-out/asmap
Retrying failed fuzz targets sequentially 1
INFO: performing bad build checks for /tmp/not-out/process_message_filterclear
Broken fuzz targets 1
('/tmp/not-out/process_message_filterclear', CompletedProcess(args=['bad_build_check', '/tmp/not-out/process_message_filterclear'], returncode=1, stdout=b'BAD BUILD: fuzzing /tmp/not-out/process_message_filterclear with afl-fuzz failed.\nafl++ setup:\nAFL_DRIVER_DONT_DEFER=1\nAFL_SKIP_CPUFREQ=1\nAFL_NO_AFFINITY=1\nAFL_FAST_CAL=1\nAFL_I_DONT_CARE_ABOUT_MISSING_CRASHES=1\nAFL_FUZZER_ARGS=-m none -x /tmp/not-out/afl++.dict -t 5000+\nAFL_NO_UI=1\nAFL_ENABLE_DICTIONARY=1\nAFL_LLVM_DICT2FILE=/out/afl++.dict\nAFL_LLVM_INSTRUMENT=CLASSIC,CTX-2\nAFL_QUIET=1\nAFL_LLVM_MODE_WORKAROUND=0\n/tmp/not-out/afl-fuzz -m none -x /tmp/not-out/afl++.dict -t 5000+ -i /tmp/process_message_filterclear_corpus -o /tmp/process_message_filterclear_afl_address_out -- /tmp/not-out/process_message_filterclear\n\x1b[1;93m[!] \x1b[1;97mWARNING: \x1b[0mMistyped AFL environment variable: AFL_DRIVER_DONT_DEFER=1\x1b[0m\n\x1b[1;92m[+] \x1b[0mLoaded environment variable AFL_SKIP_CPUFREQ with value 1\x1b[0m\n\x1b[1;92m[+] \x1b[0mLoaded environment variable AFL_NO_AFFINITY with value 1\x1b[0m\n\x1b[1;92m[+] \x1b[0mLoaded environment variable AFL_I_DONT_CARE_ABOUT_MISSING_CRASHES with value 1\x1b[0m\n\x1b[1;92m[+] \x1b[0mLoaded environment variable AFL_NO_UI with value 1\x1b[0m\n\x1b[0;36mafl-fuzz++3.13a\x1b[0m based on afl by Michal Zalewski and a large online community\n\x1b[1;92m[+] \x1b[0mafl++ is maintained by Marc "van Hauser" Heuse, Heiko "hexcoder" Ei\xc3\x9ffeldt, Andrea Fioraldi and Dominik Maier\x1b[0m\n\x1b[1;92m[+] \x1b[0mafl++ is open source, get it at https://github.com/AFLplusplus/AFLplusplus\x1b[0m\n\x1b[1;92m[+] \x1b[0mNOTE: This is v3.x which changes defaults and behaviours - see README.md\x1b[0m\n\x1b[1;92m[+] \x1b[0mLoaded environment variable ASAN_OPTIONS with value alloc_dealloc_mismatch=0:allocator_may_return_null=1:allocator_release_to_os_interval_ms=500:check_malloc_usable_size=0:detect_container_overflow=1:detect_odr_violation=0:detect_leaks=1:detect_stack_use_after_return=1:fast_unwind_on_fatal=0:handle_abort=1:handle_segv=1:handle_sigill=1:max_uar_stack_size_log=16:print_scariness=1:quarantine_size_mb=10:strict_memcmp=1:strip_path_prefix=/workspace/:symbolize=1:use_sigaltstack=1:dedup_token_length=3:abort_on_error=1:symbolize=0:detect_odr_violation=0:\x1b[0m\n\x1b[1;92m[+] \x1b[0mLoaded environment variable MSAN_OPTIONS with value print_stats=1:strip_path_prefix=/workspace/:symbolize=1:dedup_token_length=3:exit_code=86:symbolize=0\x1b[0m\n\x1b[1;92m[+] \x1b[0mNo -M/-S set, autoconfiguring for "-S default"\x1b[0m\n\x1b[1;94m[*] \x1b[0mGetting to work...\x1b[0m\n\x1b[1;92m[+] \x1b[0mUsing exponential power schedule (FAST)\x1b[0m\n\x1b[1;92m[+] \x1b[0mLoaded environment variable AFL_FAST_CAL with value 1\x1b[0m\n\x1b[1;92m[+] \x1b[0mEnabled testcache with 50 MB\x1b[0m\n\x1b[1;92m[+] \x1b[0mDisabling the UI because AFL_NO_UI is set.\x1b[0m\n\x1b[1;94m[*] \x1b[0mChecking core_pattern...\x1b[0m\n\x1b[1;92m[+] \x1b[0mYou have 16 CPU cores and 5 runnable tasks (utilization: 31%).\x1b[0m\n\x1b[1;92m[+] \x1b[0mTry parallel jobs - see docs/parallel_fuzzing.md.\x1b[0m\n\x1b[1;94m[*] \x1b[0mSetting up output directories...\x1b[0m\n\x1b[1;93m[!] \x1b[1;97mWARNING: \x1b[0mNot binding to a CPU core (AFL_NO_AFFINITY set).\x1b[0m\n\x1b[1;94m[*] \x1b[0mScanning \'/tmp/process_message_filterclear_corpus\'...\x1b[0m\n\x1b[1;92m[+] \x1b[0mLoaded a total of 1 seeds.\x1b[0m\n\x1b[1;94m[*] \x1b[0mCreating hard links for all input files...\x1b[0m\n\x1b[1;94m[*] \x1b[0mValidating target binary...\x1b[0m\n\x1b[1;92m[+] \x1b[0m\x1b[1;95mPersistent mode binary detected.\x1b[0m\n\x1b[1;92m[+] \x1b[0m\x1b[1;95mDeferred forkserver binary detected.\x1b[0m\n\x1b[1;94m[*] \x1b[0mSpinning up the fork server...\x1b[0m\n\x1b[1;92m[+] \x1b[0mAll right - fork server is up.\x1b[0m\n\x1b[1;94m[*] \x1b[0mUsing SHARED MEMORY FUZZING feature.\x1b[0m\n\x1b[1;94m[*] \x1b[0mTarget map size: 65536\x1b[0m\n\x1b[1;94m[*] \x1b[0mNo auto-generated dictionary tokens to reuse.\x1b[0m\n\x1b[1;94m[*] \x1b[0mLoading extra dictionary from \'/tmp/not-out/afl++.dict\' (level 0)...\x1b[0m\n\x1b[1;94m[*] \x1b[0mLoaded 1975 extra tokens, size range 2 B to 16 B.\x1b[0m\n\x1b[1;93m[!] \x1b[1;97mWARNING: \x1b[0mMore than 256 tokens - will use them probabilistically.\x1b[0m\n\x1b[1;92m[+] \x1b[0mLoaded a total of 338 extras.\x1b[0m\n\x1b[1;94m[*] \x1b[0mAttempting dry run with \'id:000000,time:0,orig:input\'...\x1b[0m\n\x1b[1;93m[!] \x1b[1;97mWARNING: \x1b[0mTest case results in a timeout (skipping)\x1b[0m\n\x0f\x1b)B\x1b[?25h\x1b[0m\x1b[1;91m\n[-] PROGRAM ABORT : \x1b[0mAll test cases time out, giving up!\x1b[1;91m\n         Location : \x1b[0mperform_dry_run(), src/afl-fuzz-init.c:1123\n\n', stderr=b''))
BAD BUILD: fuzzing /tmp/not-out/process_message_filterclear with afl-fuzz failed.
afl++ setup:
AFL_DRIVER_DONT_DEFER=1
AFL_SKIP_CPUFREQ=1
AFL_NO_AFFINITY=1
AFL_FAST_CAL=1
AFL_I_DONT_CARE_ABOUT_MISSING_CRASHES=1
AFL_FUZZER_ARGS=-m none -x /tmp/not-out/afl++.dict -t 5000+
AFL_NO_UI=1
AFL_ENABLE_DICTIONARY=1
AFL_LLVM_DICT2FILE=/out/afl++.dict
AFL_LLVM_INSTRUMENT=CLASSIC,CTX-2
AFL_QUIET=1
AFL_LLVM_MODE_WORKAROUND=0
/tmp/not-out/afl-fuzz -m none -x /tmp/not-out/afl++.dict -t 5000+ -i /tmp/process_message_filterclear_corpus -o /tmp/process_message_filterclear_afl_address_out -- /tmp/not-out/process_message_filterclear
[!] WARNING: Mistyped AFL environment variable: AFL_DRIVER_DONT_DEFER=1
[+] Loaded environment variable AFL_SKIP_CPUFREQ with value 1
[+] Loaded environment variable AFL_NO_AFFINITY with value 1
[+] Loaded environment variable AFL_I_DONT_CARE_ABOUT_MISSING_CRASHES with value 1
[+] Loaded environment variable AFL_NO_UI with value 1
afl-fuzz++3.13a based on afl by Michal Zalewski and a large online community
[+] afl++ is maintained by Marc "van Hauser" Heuse, Heiko "hexcoder" Eißfeldt, Andrea Fioraldi and Dominik Maier
[+] afl++ is open source, get it at https://github.com/AFLplusplus/AFLplusplus
[+] NOTE: This is v3.x which changes defaults and behaviours - see README.md
[+] Loaded environment variable ASAN_OPTIONS with value alloc_dealloc_mismatch=0:allocator_may_return_null=1:allocator_release_to_os_interval_ms=500:check_malloc_usable_size=0:detect_container_overflow=1:detect_odr_violation=0:detect_leaks=1:detect_stack_use_after_return=1:fast_unwind_on_fatal=0:handle_abort=1:handle_segv=1:handle_sigill=1:max_uar_stack_size_log=16:print_scariness=1:quarantine_size_mb=10:strict_memcmp=1:strip_path_prefix=/workspace/:symbolize=1:use_sigaltstack=1:dedup_token_length=3:abort_on_error=1:symbolize=0:detect_odr_violation=0:
[+] Loaded environment variable MSAN_OPTIONS with value print_stats=1:strip_path_prefix=/workspace/:symbolize=1:dedup_token_length=3:exit_code=86:symbolize=0
[+] No -M/-S set, autoconfiguring for "-S default"
[*] Getting to work...
[+] Using exponential power schedule (FAST)
[+] Loaded environment variable AFL_FAST_CAL with value 1
[+] Enabled testcache with 50 MB
[+] Disabling the UI because AFL_NO_UI is set.
[*] Checking core_pattern...
[+] You have 16 CPU cores and 1 runnable tasks (utilization: 6%).
[+] Try parallel jobs - see docs/parallel_fuzzing.md.
[*] Setting up output directories...
[!] WARNING: Not binding to a CPU core (AFL_NO_AFFINITY set).
[*] Scanning '/tmp/process_message_filterclear_corpus'...
[+] Loaded a total of 1 seeds.
[*] Creating hard links for all input files...
[*] Validating target binary...
[+] Persistent mode binary detected.
[+] Deferred forkserver binary detected.
[*] Spinning up the fork server...
[+] All right - fork server is up.
[*] Using SHARED MEMORY FUZZING feature.
[*] Target map size: 65536
[*] No auto-generated dictionary tokens to reuse.
[*] Loading extra dictionary from '/tmp/not-out/afl++.dict' (level 0)...
[*] Loaded 1975 extra tokens, size range 2 B to 16 B.
[!] WARNING: More than 256 tokens - will use them probabilistically.
[+] Loaded a total of 338 extras.
[*] Attempting dry run with 'id:000000,time:0,orig:input'...
[!] WARNING: Test case results in a timeout (skipping)

[-] PROGRAM ABORT : All test cases time out, giving up!
         Location : perform_dry_run(), src/afl-fuzz-init.c:1123


ERROR: 20.0% of fuzz targets seem to be broken. See the list above for a detailed information.
Check build failed.
```

</details>



With commit 9b40e02d28bbb5c392d72ca18b20043110e84f8e merged in, it is printing the exact same error. That means I am doing something wrong here. I simply called `python infra/helper.py build_fuzzers --sanitizer address  --engine afl $PROJECT_NAME && python infra/helper.py check_build  --sanitizer address  --engine afl $PROJECT_NAME` on a fresh VM.

**Comment by inferno-chromium** [comment]
> Sorry, I am not sure how to test [infra] changes. The issue on current master is:
> 
> afl timeout on master
> With commit [9b40e02](https://github.com/google/oss-fuzz/commit/9b40e02d28bbb5c392d72ca18b20043110e84f8e) merged in, it is printing the exact same error. That means I am doing something wrong here. I simply called `python infra/helper.py build_fuzzers --sanitizer address --engine afl $PROJECT_NAME && python infra/helper.py check_build --sanitizer address --engine afl $PROJECT_NAME` on a fresh VM.

since this modifies base-builder and base-runner image, before running those commands you have build those images first
python infra/helper.py build_image base-builder
python infra/helper.py build_image base-runner

let us know how that goes.
also @jonathanmetzman for review.

**Comment by MarcoFalke** [comment]
Tested 9b40e02d28bbb5c392d72ca18b20043110e84f8e and it is passing. Thanks!

```
Running: docker run --rm --privileged -i -e FUZZING_ENGINE=afl -e SANITIZER=address -e ARCHITECTURE=x86_64 -e FUZZING_LANGUAGE=c++ -v /root/oss-fuzz/build/out/bitcoin-core:/out -t gcr.io/oss-fuzz-base/base-runner test_all.py
INFO: performing bad build checks for /tmp/not-out/process_message_filterclear
INFO: performing bad build checks for /tmp/not-out/cryptofuzz-bitcoin-cryptography-w15-p4
INFO: performing bad build checks for /tmp/not-out/cryptofuzz-bitcoin-cryptography-w24-p8
INFO: performing bad build checks for /tmp/not-out/cryptofuzz-bitcoin-cryptography-w2-p2
INFO: performing bad build checks for /tmp/not-out/asmap
Check build passed.
```

**Review by MarcoFalke — APPROVED** [review]

File: infra/base-images/base-runner/bad_build_check [hunk 1]
```diff
@@ -136,10 +133,7 @@ function check_startup_crash {
     SKIP_SEED_CORPUS=1 run_fuzzer $FUZZER_NAME -seed=1337 -runs=$MIN_NUMBER_OF_RUNS &>$FUZZER_OUTPUT
     CHECK_PASSED=$(egrep "Done $MIN_NUMBER_OF_RUNS runs" -c $FUZZER_OUTPUT)
   elif [[ "$FUZZING_ENGINE" = afl ]]; then
-    # TODO(https://github.com/google/oss-fuzz/issues/2470): Dont use
-    # AFL_DRIVER_DONT_DEFER by default, support .options files in
-    # bad_build_check instead.
-    AFL_DRIVER_DONT_DEFER=1 AFL_NO_UI=1 SKIP_SEED_CORPUS=1 timeout --preserve-status -s INT 20s run_fuzzer $FUZZER_NAME &>$FUZZER_OUTPUT
+    AFL_FORKSRV_INIT_TMOUT=30000 AFL_NO_UI=1 SKIP_SEED_CORPUS=1 timeout --preserve-status -s INT 35s run_fuzzer $FUZZER_NAME &>$FUZZER_OUTPUT
```
Comment (jonathanmetzman) [inline infra/base-images/base-runner/bad_build_check:136]: Are we doing this on ClusterFuzz?
Or should we change things?

**Review by jonathanmetzman — APPROVED** [review]
LGTM

File: infra/base-images/base-runner/bad_build_check (hunk 1 above)
Comment (vanhauser-thc) [reply infra/base-images/base-runner/bad_build_check:136]: AFL_DRIVER_DONT_DEFER is in src/python/tests/core/bot/fuzzers/afl/afl_launcher_test.py --- you know my problems running & debugging clusterfuzz so I didnt touch that. (as long as nobody complains it is not working ...)
AFL_FORKSRV_INIT_TMOUT is something that should be added to clusterfuzz I think. I can do a PR tomorrow

Comment (vanhauser-thc) [reply infra/base-images/base-runner/bad_build_check:136]: ah the AFL_DRIVER_DONT_DEFER is just in a test, so doesnt apply.
PR for forkserver init timeout is in.

**Pull request merged**